velox/functions/lib/Map.h

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/CheckDuplicateKeys.h"
+
+namespace facebook::velox::functions {
+
+static const char* kNullKeyErrorMessage = "map key cannot be null";
+static const char* kIndeterminateKeyErrorMessage =
+    "map key cannot be indeterminate";
+
+template <bool AllowDuplicateKeys>
+class MapFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VELOX_CHECK_EQ(args.size(), 2);
+
+    auto keys = args[0];
+    auto values = args[1];
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto decodedKeys = decodedArgs.at(0);
+    auto decodedValues = decodedArgs.at(1);
+
+    static const char* kArrayLengthsMismatch =
+        "Key and value arrays must be the same length";
+    auto checkNullsInKey =
+        [&](const auto& keysElements, auto offset, auto size) {
+          for (auto i = 0; i < size; ++i) {
+            VELOX_USER_CHECK(
+                !keysElements->isNullAt(offset + i), kNullKeyErrorMessage);
+
+            VELOX_USER_CHECK(
+                !keysElements->containsNullAt(offset + i),
```

Comment:
Can we move the function to Map.cpp as this https://github.com/facebookincubator/velox/blob/main/velox/functions/lib/MapConcat.h